deploy: add API for rbd/CSIDriver and rbd/ceph-csi-config ConfigMap

nixpanic opened the pull request:
This is similar to #2551 when OpenShift SecurityContextConstraints are exposed over an API.

The rbd/CSIDriver  and rbd/ceph-csi-config ConfigMap are deployment artifacts that should be made available when Ceph-CSI is deployed. Automated deployers (like Rook) can consume this API and remove any redundant maintained deployment files in the future.

Updates: #2542

---

<details>
<summary>Show available bot commands</summary>

These commands are normally not required, but in case of issues, leave any of
the following bot commands in an otherwise empty comment in this PR:

- `/retest ci/centos/<job-name>`: retest the `<job-name>` after unrelated
  failure (please report the failure too!)
- `/retest all`: run this in case the CentOS CI failed to start/report any test
  progress or results

</details>

nixpanic committed:
deploy: move rbd/CSIDriver to API

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic committed:
deploy: move rbd/ceph-csi-config ConfigMap to API

Signed-off-by: Niels de Vos <ndevos@redhat.com>

Madhu-1 commented on tools/yamlgen/main.go:
```diff
@@ -35,18 +37,25 @@ const header = `---
 		"../deploy/scc.yaml",
-		ocp.NewSecurityContextConstraintsYAML,
-		ocp.SecurityContextConstraintsDefaults,
+		reflect.ValueOf(ocp.NewSecurityContextConstraintsYAML),
+		reflect.ValueOf(ocp.SecurityContextConstraintsDefaults),
+	},
+	{
+		"../deploy/rbd/kubernetes/csidriver.yaml",
+		reflect.ValueOf(rbd.NewCSIDriverYAML),
+		reflect.ValueOf(rbd.CSIDriverDefaults),
+	},
+	{
+		"../deploy/rbd/kubernetes/csi-config-map.yaml",
+		reflect.ValueOf(rbd.NewCSIConfigMapYAML),
+		reflect.ValueOf(rbd.CSIConfigMapDefaults),
```
Maybe we need to expose Register API instead of adding more items to the list? everyone can register in their own package we just need to generate templates for Registered one?

nixpanic replied:
Yes, maybe. That can be done in addition to what is proposed here. Not sure if that is needed to have for the initial version of the API?

nixpanic committed:
build: remove unneeded empty YAML document from deployment artifacts

The generated files under the deploy/ directory contain an empty YAML
document that may cause confusion for some versions of kubectl. Dropping
the unneeded `---` start of the file for the header should make parsing
of the deployment artifacts a little less error prone.

Signed-off-by: Niels de Vos <ndevos@redhat.com>

nixpanic commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

nixpanic commented:
> /retest ci/centos/mini-e2e-helm/k8s-1.21

Failed with #2264

nixpanic commented:
/retest ci/centos/mini-e2e-helm/k8s-1.20

nixpanic commented:
/retest ci/centos/mini-e2e-helm/k8s-1.21

nixpanic commented:
> /retest ci/centos/mini-e2e-helm/k8s-1.20

Tests timed out after 90 minutes ([logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e-helm_k8s-1.20/detail/mini-e2e-helm_k8s-1.20/2774/pipeline/))

nixpanic commented:
/retest ci/centos/mini-e2e/k8s-1.20

nixpanic commented:
> /retest ci/centos/mini-e2e/k8s-1.20

Failed with #2204 ([logs](https://jenkins-ceph-csi.apps.ocp.ci.centos.org/blue/organizations/jenkins/mini-e2e_k8s-1.20/detail/mini-e2e_k8s-1.20/2788/pipeline))

nixpanic commented:
@mergifyio rebase
